components/script/dom/readablestream.rs

```diff
@@ -423,6 +431,47 @@ impl ReadableStream {
 
         // TODO: react to sourceCancelPromise.
     }
+
+    /// <https://streams.spec.whatwg.org/#readable-stream-reader-generic-release>
+    pub fn release_lock(
+        &self,
+        closed_promise: &Promise,
+        read_requests: DomRefCell<VecDeque<ReadRequest>>,
+    ) {
+        // step 2.1.3
+        assert!(self.is_locked());
+
+        if self.is_readable() {
+            // step 2.1.4
+            closed_promise.reject_error(Error::Type("stream state is not readable".to_owned()));
+        } else {
+            // step 2.1.5
+            closed_promise.reject_error(Error::Type(
+                "Cannot release lock due to stream state.".to_owned(),
+            ));
+        }
+
+        // step 2.1.6
+        closed_promise.set_promise_is_handled();
+
+        // step 2.1.7
+        self.perform_release_steps();
+
+        // step 2.1.8 & 2.1.9
+        self.set(None);
+
+        // step  3.1
+        let error = Error::Type("No chunks are available because the stream is errored".to_owned());
```

Comment:
This can be moved to a method implementing https://streams.spec.whatwg.org/#abstract-opdef-readablestreamdefaultreadererrorreadrequests on the reader.  You can use the code at https://streams.spec.whatwg.org/#readable-stream-error and move it to a new method called in both places.